data_safe_haven/commands/base_command.py

```diff
@@ -0,0 +1,38 @@
+# Standard library imports
+from abc import ABC, abstractmethod
+from typing import Any
+
+# Third-party imports
+import typer
+
+# Local imports
+from data_safe_haven.utility import Logger
+
+
+class IBaseCommand(ABC):
+    """Command with entrypoint for use with Typer"""
+
+    @abstractmethod
+    def entrypoint(self, *args: Any, **kwargs: Any) -> None:
+        """Typer command line entrypoint"""
+        pass
+
+
+class BaseCommand(IBaseCommand):
+    """IBaseCommand with logger"""
+
+    def __init__(self):
+        """Constructor"""
+        self.logger = Logger()
+
+
+class CommandGroup(typer.Typer):
+    """Command group containing subcommands"""
+
+    def subgroup(self, cls: type["CommandGroup"], name: str, help: str) -> None:
+        """Register a command group"""
+        self.add_typer(cls(), name=name, help=help)
+
+    def subcommand(self, cls: type[BaseCommand], name: str, help: str) -> None:
+        """Register a command"""
+        self.command(name=name, help=help)(cls().entrypoint)
```

Comment:
Is this because the subcommands are custom classes rather than `Typer.Typer` instances? Is that necessary?

What I'm imagining is more like the Typer examples,

in `__main.py__` (or whatever our entrypoint is)

```python
import teardown
import admin

app = Typer.Typer()

app.add_typer(teardown.app, name='teardown')
app.add_typer(admin.app, name='admin')
```

and in `admin.py`

```python
app = Typer.Typer()

@app.command()
def add_user(...):
  ...
```

It is more idiomatic and will be easier to understand for people who have used Typer.

Reply:
That part is fine, but not all of the top-level commands are groups of commands, so you'll still need something like

```
import teardown
import admin
from initialise_command import entrypoint
from data_safe_haven.functions import validate_aad_guid

app = Typer.Typer()

app.add_typer(teardown.app, name='teardown')
app.add_typer(admin.app, name='admin')

@app.command(help="Initialise a Data Safe Haven deployment.")
def init(admin_group: Annotated[
            Optional[str],
            typer.Option(
                "--admin-group",
                "-a",
                help="The ID of an Azure group containing all administrators.",
                callback=validate_aad_guid,
            ),
        ] = None,
        location: Annotated[
            Optional[str],
            typer.Option(
                "--location",
                "-l",
                help="The Azure location to deploy resources into.",
            ),
        ] = None,
        name: Annotated[
            Optional[str],
            typer.Option(
                "--deployment-name",
                "-d",
                help="The name to give this Data Safe Haven deployment.",
                callback=validate_aad_guid,
            ),
        ] = None,
        subscription: Annotated[
            Optional[str],
            typer.Option(
                "--subscription",
                "-s",
                help="The name of an Azure subscription to deploy resources into.",
            ),
        ] = None,):
    entrypoint(admin_group, location, name, subscription)
```

which I did start off with, but I thought it made things messier than necessary.

Reply:
I (strongly) prefer this, the hierarchy is clear.

Could always put `def init(...)` in another file and import the `typer.Typer` instance there if it is about keeping files small and 'on topic'.